src/main/java/com/mongodb/kafka/connect/source/MongoSourceTask.java

```diff
@@ -309,26 +333,32 @@ private void setCachedResultAndResumeToken() {
             cursor = createCursor(sourceConfig, mongoClient);
         }
 
-        try {
-            BsonDocument next = cursor.tryNext();
-            if (next == null && cursor.getServerCursor() == null) {
-                cursor.close();
-                cursor = null;
+        if (cursor != null) {
+            try {
+                BsonDocument next = cursor.tryNext();
+                // The cursor has been closed by the server
+                if (next == null && cursor.getServerCursor() == null) {
+                    invalidatedCursor = true;
+                    cursor.close();
+                    cursor = createCursor(sourceConfig, mongoClient);
```

Comment:
Should this be preceded by

     cursor = null;

so that it's null on line 347 on an exception?

Reply:
Done